# Student Performance Indicator

## Life cycle of Machine learning Project
- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

## 1. Understanding the Problem Statement
- This project understands how the student's performace(test scores) is affected by the variables such as Gender, Ethnicity Parental level of education, Lunch and Test preparation course.

## 2. Data Collection
Dataset Source: https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977

# 2.1 Import data and required packages
### Importing Pandas, Numpy, Matplotlib and seaborn

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

### Import CSV data as Pandas DataFrame

In [3]:
df=pd.read_csv("data/stud.csv")

#### Show top 5

In [4]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

#### Shape of the dataset

In [5]:
df.shape

(1000, 8)

## 2.2 Dataset information

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# 3. Data Checks to perform
- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

### 3.1 Check missing values

In [7]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

#### There is no missing values

### 3.2 Check duplicates

In [8]:
df.duplicated().sum()

0

#### There is no duplicate values

### 3.3 Check Date Type

In [9]:
df.dtypes

gender                         object
race_ethnicity                 object
parental_level_of_education    object
lunch                          object
test_preparation_course        object
math_score                      int64
reading_score                   int64
writing_score                   int64
dtype: object

### 3.4 Check number of Unique values in each column

In [10]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

### 3.5 Check statistics of data set

In [11]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

Insight
- From the description of numerical data, all means are very close to each other - between 66 and 69.16
- All the standard deviations are also close to each other - between 14.60 and 15.19
- There is a minimum score 0 for math, for writng minimum is much higher which is 10, for reading minimum is 17

#### 3.5 Check various categories present in the different categorical column

#### Exploring data
separating numerical and categorical columns

In [12]:
numerical_features=[i for i in df if df[i].dtypes != "O"]
print(numerical_features)
categorical_features=[i for i in df if df[i].dtypes == "O"]
print(categorical_features)

['math_score', 'reading_score', 'writing_score']
['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [13]:
print("we have {} numerical features {}".format(len(numerical_features),numerical_features))
print("we have {} categorical features {}".format(len(categorical_features),categorical_features))

we have 3 numerical features ['math_score', 'reading_score', 'writing_score']
we have 5 categorical features ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [14]:
print("categories in gender: {} ".format(df["gender"].unique()))
print("categories in race ethnicity: {} ".format(df["race_ethnicity"].unique()))
print("categories in parental level of education: {} ".format(df["parental_level_of_education"].unique()))
print("categories in lunch: {} ".format(df["lunch"].unique()))
print("categories in test preparation course: {} ".format(df["test_preparation_course"].unique()))

categories in gender: ['female' 'male'] 
categories in race ethnicity: ['group B' 'group C' 'group A' 'group D' 'group E'] 
categories in parental level of education: ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school'] 
categories in lunch: ['standard' 'free/reduced'] 
categories in test preparation course: ['none' 'completed'] 


## Adding columns for "Total Score" and "Average"

In [15]:
df["total score"]=df["math_score"]+df["reading_score"]+df["writing_score"]
df["average"]=df["total score"]/3

In [16]:
df.head(10)

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   
5                    none          71             83             78   
6               completed          88             95             92   
7                    none          40             43             39   
8               completed          64             64             67   
9                    none          38             60             50   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333  
5          232  77.333333  
6          275  91.666667  
7          122  40.666667  
8          195  65.000000  
9          148  49.333333

In [17]:
full_in_math = df[df['math_score']==100]['math_score'].count()
full_in_reading = df[df['reading_score']==100]['reading_score'].count()
full_in_writing = df[df['writing_score']==100]['writing_score'].count()

print(f'Number of students scored full marks in maths {full_in_math}')
print(f'Number of students scored full in reading {full_in_reading}')
print(f'Number of students scored full in writing {full_in_writing}')

Number of students scored full marks in maths 7
Number of students scored full in reading 17
Number of students scored full in writing 14


In [18]:
math_less_20 = df[df['math_score'] <= 20]['math_score'].count()
reading_less_20 = df[df['reading_score'] <= 20]['reading_score'].count()
writing_less_20 = df[df['writing_score'] <= 20]['writing_score'].count()

print(f'Number of students with less than 2 marks in maths: {math_less_20}')
print(f'Number of students with less than 20 marks in reading: {reading_less_20}')
print(f'Number of students with less than 20 marks in writing: {writing_less_20}')

Number of students with less than 2 marks in maths: 4
Number of students with less than 20 marks in reading: 1
Number of students with less than 20 marks in writing: 3


Insight

- from the above values we can conclude that students performed best in reading
- worst in maths